### DISCUSSION TITLE
[labs/router] Thoughts on the proposed router

### DISCUSSION THREAD

**Discussion by zandaqo** [lit/lit#2349] [Ideas]
I have been using lit-html for years with my own [router](https://github.com/zandaqo/compago#routing-with-routercontroller) since the existing ones were too convoluted for my taste. The new router package seems like a nice minimalist solution for the task, and I would be happy to replace my router with a "built-in" solution one day. I'd like to point out a couple things and make some subjective suggestions regarding them:

1. **Paths**
   URLPattern is nice, yet in practice I discovered that RegExp with matching groups are ample for real world use cases. Especially, with the RESTful URIs: pretty much all boil down to `collection/:id` which is not terribly complicated to write as a regex, e.g. `/^\/users\/(?<id>[0-9a-z])\/?$/`. Plus in most cases, the parameters can be "tightened" in regex avoiding false matches. The more complicated parameters/state can be supplied in search parameters and/or History API's state. Hence,
   **Suggestion**: Allow RegExps for paths. It will make URLPattern optional and give a slight performance edge (admittedly, not terribly important here).
2. **Global click listener**
   A questionable move in terms of performance: most clicks in a SPA are not navigational, yet all will incur an overhead with an extra listener. The listener itself has to impose arbitrary restrictions (like the click being inside an `<a>`) to filter the events. In my router, I went with a directive that essentially just triggers a popstate event, it does result in extra work of defining links (e.g. `<a href="users/${id}" @click=${navigate()}>${id}</a>`), but gives much more freedom in terms of used elements, events, and a possibility of supplying extra data such as state. Admittedly, it doesn't work with user provided links, but since the components where those types of links can appear are limited and known (comments, articles, etc.) those can use local click handlers.
   **Suggestion**: Make the global handler optional and maybe give another way of triggering the navigation events?
3. **Child Routes**
   Every time I used this sort of top-down control of routers it ended up unnecessarily complicating things: tying children to the global parent that does too much accounting. In the end, I made all routers to listen to the global `popstate` with "child" routers skipping the url if it doesn't start with their `root` attribute, that is, they are "scoped" to their roots. This does force you to add the child roots in the global router (usually as no-ops), but otherwise the routers are independent. Plus, the "tail matching" pattern `/*` can be used as a "fallback" route for missing pages.
   **Suggestion:** No suggestions here, just pointing out that we don't have to have this sort of hierarchies simply to have scoped routers.

**Comment by UserGalileo** [comment]
I'll add a few things that imho should be discussed! :)

## Lazy Loading
I think it'd be pretty much needed an optional `setup` callback for each route, in order to let the router import some files when a route is matched (eg. loading a component lazily):

```ts
{
  path: '/movie/:id',
  render: (params) => html`<app-movie .movieId=${params.id}></app-movie>`,
  setup: () => import('./pages/movie.component'),
}
```

This way we keep the `render` function simple and pure, without side-effects.

## Animations
Would it be possible to animate route transitions with this router? I mean: if a route changes, the previous component is immediately destroyed AFAIK, am I right?

I looked at Vaadin's Router previously (https://vaadin.com/router) and it does something interesting, keeping the old component in the DOM allowing for a CSS transition. Admittedly I've had some problems with it (you can look at the demo on that page to see that if you change fast enough the transition doesn't look right, and it messes up the lifecycle of the components, meaning that when dealing with some sort of outer state a-la Redux if I clear some state on disconnect it'll happen after the next component is connected) so I'm not sure that's the right approach. But I think it'll be useful to consider this necessity so that, eventually, it can be done in the future.

## Children routes
Would it be useful to also have children routes instead of nested routes? I mean:

```ts
{
  path: '/movie',
  children: [
    { path: '', render: ... },
    { path: 'detail', render: ... },
    { path: ':id', render: ... }
  ]
}
```

This would allow avoiding writing `/movie` in all paths, and maybe it'd be useful for some configuration to be written on the parent route only instead of every one of them. (a single setup function? guards?). 

This would be different from nested routes because you wouldn't have a nested outlet where a new container is needed. It'd simply be a convenient way for the user to write some routes, nothing more.

## Redirects
I see no redirects mentioned anywhere in the PR. I think it'd be really useful and not hard to achieve, what do you think?

```ts
{
  path: '...',
  redirectTo: '/home' // Could be relative or absolute
}
```

## Guards
What would be the preferred way to achieve some sort of Guard behavior?

Without adding anything to the router, I guess it could be achieved with a custom component which performs a redirect:

```ts
{
  path: '/movie',
  render: () => html`<auth-guard><app-movie></app-movie></auth-guard>`
}
```

But I'm not sure this is ideal, because the slotted content would always be there in the DOM and it'd be created before the guard's checks. Another solution would be to use the Render Props pattern:

```ts
{
  path: '/movie',
  render: () => html`<auth-guard .render=${() => html`<app-movie></app-movie>`}></auth-guard>`
}
```

I guess this would work better but it's different from anything I've ever written so I don't know! What do you think? Or would it be the case to add some sort of Guard configuration to the router to keep the render callbacks cleaner?

**Comment by justinfagnani** [comment]
@zandaqo thanks for the feedback. Addressing your points:

1. Paths. I think utilizing URLPattern is simply the right thing to do for a new router. We'll have to ship the URLPattern polyfill initially, but in a future version we'll be able to assume native support and have a much smaller router. So whatever URLPattern supports for matches, we'll support. URLPattern _does_ indeed support regex matches though, so your request is already met: https://developer.mozilla.org/en-US/docs/Web/API/URL_Pattern_API#regex_matchers

2. Global listeners. Currently there is no other way to intercept navigating clicks. When the [App History API](https://wicg.github.io/app-history/) lands, we'll be able to listen for the [`navigate`](https://wicg.github.io/app-history/#navigate-event) and not listen for clicks at all, but until then this is the best approach. I don't think we want to require that all `<a>` tags have a special directive added to be supported by the router. That might not always be possible anyway.

I also don't think there will be any negative performance impact. In the case of a click event that isn't relevant to the router, we do very little work in the handler. Nearly all routers work this way already and I haven't heard of performance issues.

3. Child Routes. I think automatic mounting of child routers is a nice DX and composability feature. Requiring child routers to know their mount path would mean that it's either hard coded or has to be passed down from the parent. I think we can get this right, but if not we can fall back to that type of mount point.

**Comment by zandaqo** [comment]
@justinfagnani I appreciate your time and consideration here, just to clarify few things:
> I think utilizing URLPattern is simply the right thing to do for a new router.

I concur, I'm glad the router is going with URLPattern instead of  the usual `path-to-regex`. I suggest allowing regexps alongside URLPattern to speed up the adoption for those who might not need all the power of the API and might be hesitant to add yet another polyfill.

>  When the App History API lands, we'll be able to listen for the navigate and not listen for clicks at all, but until then this is the best approach.

Somehow I missed this API, thanks for bringing it up! That would indeed solve the problem nicely.

> I also don't think there will be any negative performance impact.

I agree, I'm probably being old and paranoid here, it's not like we do hundreds of clicks per second for it to matter on the modern hardware. 

> Requiring child routers to know their mount path would mean that it's either hard coded or has to be passed down from the parent.

To avoid hard coding or passing down we can use the Context API. That is, a child router can check up the DOM tree for another, parent, router and prepend its mount point with the mount point of the parent. E.g. if we have a parent router mounted at `/collection` and put somewhere down the tree a child router mounted at `/item`, upon attaching it can get the parent's mounting point from the context and set its own mounting point to `/collection/item`. This way the routers only have to know their own segment of the path but otherwise can be re-usable.

**Reply by mark-dropbear** [reply]
Can we not just set up lazy loading the routes as follows:
```js
    {
      pattern: new URLPattern({ pathname: '/' }),
      render: () => {
        import('../pages/home-page.js');
        return html`<home-page></home-page>`;
      },
    },
```

**Comment by markcellus** [comment]
This is an interesting discussion! I've attempted to make a declarative router called [router-component](https://github.com/markcellus/router-component) that routes to a pages (aka web components). It supports nested routes, global click listeners for links, and animations. It doesn't even really depend on lit elements, you can use it with just bare web components. 

I haven't tried out the lit router just yet, but I'll take a look at it soon. Although I do wonder if this should even be a "lit router" vs a general purpose router for web components. :thinking:

**Comment by christophe-g** [comment]
I have recently [played a bit with `labs/router`](https://lit.dev/playground/#gist=db4627d8101e800c83394cab7eb6282a) with great enthusiasm and pleasure. Here are my remarks/questions.

First of all: thank you ! It looks great - I want to be using it. 

## What I like: 
- being able to declare lit-template as route and stay within  pure `lit` environment.
- having routes/templates scoped with current host
- child routes
- mutable routes 

## goto
I was actually expecting `goto` to keep location in sync with routes. The current implementation only set the current route. What about having two methods 1. `setCurrentRoute` as the current `goto`; and `goto(path, title) {history.pushState(title, null, path); this.goto(path)}`?

Having a method for both manipulating the url and the route makes sense as [history.pushState won't trigger a `popsate` event](https://developer.mozilla.org/en-US/docs/Web/API/Window/popstate_event). 

## link
I have the impression link is not working well with child routes: 
- go to the playground above 
- click home (current location: '/')
- click projects (current location: '/projects')
- click profile/1 (current location: '/projects/profile/1')
- click about (current location: '/project/profile/about')
I would expect link resolves the URL relatively to the current route, hence '/about' and not '/project/profile/about', but I might be missing something.

## child routes
1. It seems that child routes tries to resolve the path when they should not be active: 
- click home (current location: '/')
- click projects (current location: '/projects')
- click notExist 
- see error (`Error: No route found for notExist`) thrown by child route, which should be inactive

2. Path on parent routes needed to start with '/'. On child routes, the starting '/' needs to be removed so as paths to be resolved. Is it the expected behavior ?

3. I was not sure how to handle cases like: 
```js
 {
      path: '/projects',
      enter: async (params) => {
        console.info('entering projects');
        await this.router.goto('/projects/all')
        return false;
      }
    },
    {
      path: '/projects/*',
      render: () => html`<h2>Projects</h2><test-app-project></test-app-project>`,
    },
```

Ideally in this case, I would like to have: 
- a way to redirect `/projects` to `/projects/*` and
- a way to activate a default path (`/all` or `/`)
